msg_helpers.go

```diff
@@ -261,28 +261,42 @@ func packUint64(i uint64, msg []byte, off int) (off1 int, err error) {
 }
 
 func unpackString(msg []byte, off int) (string, int, error) {
-	if off+1 > len(msg) {
+	start := off + 1
+	if start > len(msg) {
 		return "", off, &Error{err: "overflow unpacking txt"}
 	}
 	l := int(msg[off])
-	if off+l+1 > len(msg) {
+	end := start + l
+	if end > len(msg) {
 		return "", off, &Error{err: "overflow unpacking txt"}
 	}
 	var s strings.Builder
-	s.Grow(l)
-	for _, b := range msg[off+1 : off+1+l] {
+	p := msg[start:end]
+	for i, b := range p {
 		switch {
 		case b == '"' || b == '\\':
+			if s.Len() == 0 {
+				s.Grow(l * 2)
```

Comment:
why `* 2 here`?

Reply:
The `* 2` is because we need more than `len(s)` bytes to store the result, but we don't know how much more space we need so I arbitrarily picked `* 2` as it provides plenty of headroom while still being cheaper than than `len(s)`, which requires at least `* 3` bytes and two allocs - see [`Builder.grow()`](https://github.com/golang/go/blob/release-branch.go1.13/src/strings/builder.go#L64-L70).

A lower value (say `* 1.5`) could be used but one challenge is that escape sequences require either 2 or 4 bytes.  `* 2` is pretty aggressive when escaping with 2 bytes, but doesn't leave a lot of headroom when escaping with 4 bytes.